Skip configuring NP related if network_policy_enforcing_mode is not set

Pavani-Panakanti opened the pull request:
**What type of PR is this?**
Enhancement

**Which issue does this PR fix?**:
We should fail on pod creation if we can't connect to np agent and fail to attach probes at pod start.

**What does this PR do / Why do we need it?**:
Some of the cx might have deleted NP container if they are not using network policies, so we are using env_var NETWORK_POLICY_ENFORCING_MODE to check if cx wants network policy configuration or not. If cx is deleting np container they have to not set this env_var. When this is not set we don't try to connect to np container at all

**Testing done on this change**:
standard mode, np enabled
strict mode, np enabled
"" mode, np enabled --> invalid mode fails
not set, np container present --> we skip reaching out to np container
not set, np container not present --> we skip call to np container 
set to some value, np container not present --> pod creates fail


**Will this PR introduce any new dependencies?**:
No

**Will this break upgrades or downgrades? Has updating a running cluster been tested?**:
No

**Does this change require updates to the CNI daemonset config files to work?**:
No

**Does this PR introduce any user-facing change?**:
Cx who have removed NP container should not set NETWORK_POLICY_ENFORCING_MODE

```release-note
Cx who have removed NP container should not set NETWORK_POLICY_ENFORCING_MODE
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jayanthvn commented on charts/aws-vpc-cni/templates/daemonset.yaml:
```diff
@@ -81,9 +81,13 @@ spec:
             timeoutSeconds: {{ .Values.readinessProbeTimeoutSeconds }}
           env:
 {{- range $key, $value := .Values.env }}
+{{- if and (eq $key "NETWORK_POLICY_ENFORCING_MODE") (not .Values.nodeAgent.enabled) }}
```
I'm fine with this but a minor nit, we could make it readable wdyt? something like this -

```
{{- $skipKey := and (eq $key "NETWORK_POLICY_ENFORCING_MODE") (not .Values.nodeAgent.enabled) }}
{{- if not $skipKey }}
  - name: {{ $key }}
    value: {{ $value | quote }}
{{- else }}
  # Skipping NETWORK_POLICY_ENFORCING_MODE because nodeAgent is disabled
{{- end }}

```

Pavani-Panakanti replied:
yeah this looks more readable. Updated

jayanthvn commented on cmd/routed-eni-cni-plugin/cni.go:
```diff
@@ -451,6 +454,10 @@ func del(args *skel.CmdArgs, cniTypes typeswrapper.CNITYPES, grpcClient grpcwrap
+	if r.NetworkPolicyMode == "" {
+		log.Infof("NETWORK_POLICY_ENFORCING_MODE is not set")
+		return nil
+	}
```
In line 466, we need the same fix as add for logs right?

```
		log.Errorf("Failed to connect to network policy agent: %v", err)
		return errors.New("add cmd: failed to setup network policy")
```

Pavani-Panakanti replied:
We were never failing in NP when there are any errors in delete path so followed the same here but it might be better to make that more strict to avoid leaving any maps behind. Updated

jayanthvn commented:
Can you check metrics helper compilation is failing?

Also can you add a note here - https://github.com/aws/amazon-vpc-cni-k8s/tree/master?tab=readme-ov-file#network_policy_enforcing_mode-v1171 for use case of someone removing the NP container?

Pavani-Panakanti commented:
Fixed metrics.go and also updated README. We still have one vulncheck failing for updating goversion to 1.24.2 but we don't have the base image for 1.24.2 yet

oliviassss commented on pkg/ipamd/ipamd.go:
```diff
@@ -1846,13 +1845,14 @@ func EnablePodENI() bool {
+		return value, nil
 	}
-	return defaultNetworkPolicyMode, nil
+	return "", errors.New("invalid Network policy mode, supported modes: none, strict, standard")
```
just to confirm backward compatibility, but default the helm chart has the `NETWORK_POLICY_ENFORCING_MODE=strict` right? So the change should not break any existing cx behavior

Pavani-Panakanti replied:
By default network_policy_enforcing_mode is set to standard https://github.com/aws/amazon-vpc-cni-k8s/blob/master/charts/aws-vpc-cni/values.yaml#L90. It should not break any existing behavior. Cx who explicitly remove the env_var entirely will be impacted and we upgraded doc to state that

jaydeokar reviewed (approved):
lgtm
